src/cargo/core/global_cache_tracker.rs

```diff
@@ -0,0 +1,1754 @@
+//! try to not impact it too much. One part that has a performance concern is
+//! that the clean command will synchronize the database with whatever is on
+//! disk if needed (in case files were added by older versions of cargo that
+//! don't do cache tracking, or if the user manually deleted some files). This
+//! can potentially be very slow, especially if the two are very out of sync.
+//!
+//! ## Filesystems
+//!
+//! Everything here is sensitive to the kind of filesystem it is running on.
+//! People tend to run cargo in all sorts of strange environments that have
+//! limited capabilities, or on things like read-only mounts. The code here
+//! needs to gracefully handle as many situations as possible.
+//!
+//! The sections above about performance and locking are very relevant when
+//! considering different filesystems.
+//!
+//! There are checks for read-only filesystems, which is generally ignored.
+
+use crate::core::gc::GcOpts;
+use crate::core::Verbosity;
+use crate::ops::CleanContext;
+use crate::util::cache_lock::CacheLockMode;
+use crate::util::interning::InternedString;
+use crate::util::sqlite::{self, basic_migration, Migration};
+use crate::util::{Filesystem, Progress, ProgressStyle};
+use crate::{CargoResult, Config};
+use anyhow::{bail, Context};
+use cargo_util::paths;
+use rusqlite::{params, Connection, ErrorCode};
+use std::collections::{hash_map, HashMap};
+use std::path::{Path, PathBuf};
+use std::time::{Duration, SystemTime};
+use tracing::{debug, trace};
+
+/// The filename of the database.
+const GLOBAL_CACHE_FILENAME: &str = ".global-cache";
+
+const REGISTRY_INDEX_TABLE: &str = "registry_index";
+const REGISTRY_CRATE_TABLE: &str = "registry_crate";
+const REGISTRY_SRC_TABLE: &str = "registry_src";
+const GIT_DB_TABLE: &str = "git_db";
+const GIT_CO_TABLE: &str = "git_checkout";
+
+/// How often timestamps will be updated.
+///
+/// As an optimization timestamps are not updated unless they are older than
+/// the given number of seconds. This helps reduce the amount of disk I/O when
+/// running cargo multiple times within a short window.
+const UPDATE_RESOLUTION: u64 = 60 * 5;
+
+/// Type for timestamps as stored in the database.
+///
+/// These are seconds since the Unix epoch.
+type Timestamp = u64;
+
+/// Tracking for the global shared cache (registry files, etc.).
+#[derive(Debug)]
+pub struct GlobalCacheTracker {
+    /// Connection to the SQLite database.
+    conn: Connection,
+    auto_gc_checked_this_session: bool,
+}
+
+/// This is a cache of modifications that will be saved to disk all at once
+/// via the [`DeferredGlobalLastUse::save`] method.
+///
+/// This is here to improve performance.
+#[derive(Debug)]
+pub struct DeferredGlobalLastUse {
+    /// Cache of registry keys, used for faster fetching.
+    ///
+    /// The key is the registry name (which is its directory name) and the
+    /// value is the `id` in the `registry_index` table.
+    registry_keys: HashMap<InternedString, i64>,
```

Comment:
Can we use a new type for the table id's?  I feel like it'd make the intent of the code clearer and likely reduce the chance for bugs